cargo-miri/bin.rs

```diff
@@ -596,16 +596,13 @@ fn phase_cargo_rustc(args: env::Args) {
     let target_crate = is_target_crate();
     let print = get_arg_flag_value("--print").is_some(); // whether this is cargo passing `--print` to get some infos
 
-    // rlib and cdylib are just skipped, we cannot interpret them and do not need them
+    // cdylib is just skipped, we cannot interpret it and do not need it
     // for the rest of the build either.
-    match get_arg_flag_value("--crate-type").as_deref() {
-        Some("rlib") | Some("cdylib") => {
-            if verbose {
-                eprint!("[cargo-miri rustc] (rlib/cdylib skipped)");
-            }
-            return;
+    if get_arg_flag_value("--crate-type").as_deref() == Some("cdylib") {
+        if verbose {
+            eprint!("[cargo-miri rustc] (cdylib skipped)");
         }
-        _ => {},
+        return;
```

Comment:
I found that skipping the crate if `get_arg_flag_value("--crate-type")` is something is probably wrong, because Cargo passes *multiple* `--crate-type` to the compiler.

If I add
```rust
issue_1567::use_the_dependency();
```
to `test-cargo-miri/src/lib.rs`, the test will fail:
```
## Running `cargo miri` tests
A libstd for Miri is now available in `/my/home/directory/.cache/miri/HOST`.
Testing `cargo miri run` (no isolation)...
--- BEGIN stdout ---
--- END stdout ---
--- BEGIN stderr ---
error: extern location for issue_1567 does not exist: /my/home/directory/miri/test-cargo-miri/target/x86_64-unknown-linux-gnu/debug/deps/libissue_1567.rmeta
 --> src/lib.rs:6:5
  |
6 |     issue_1567::use_the_dependency();
  |     ^^^^^^^^^^

error: aborting due to previous error

error: could not compile `cargo-miri-test`

To learn more, run the command again with --verbose.
--- END stderr ---

TEST FAIL: exit code was 101
```
I think that is because Cargo passes *both* `--crate-type cdylib` and `--crate-type rlib` to the compiler and only invokes the compiler once. If `cargo-miri` skips the crate here, the `rlib` won't be produced.

Maybe just dropping `--crate-type cdylib` and continuing to compile with only `--crate-type rlib` (or skip the compilation if there's no more `--crate-type`) makes more sense.

Reply:
> (EDIT: Hit #1705 when experimenting with that... Trying to fix that together...)

So I suggest we tackle that in a future MR. Good catch though, this should be added to the test suite then (together with a `crate-type = ["lib", "staticlib", "cdylib"]` crate that triggers [this](https://github.com/rust-lang/miri/issues/1705#issuecomment-778651455) behavior).